File: python/Diameter of Binary Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root):
        self.ans = 0

        def depth(p):
            if not p:
                return 0
            left, right = depth(p.left), depth(p.right)
            self.ans = max(self.ans, left + right)
            return 1 + max(left, right)

        depth(root)
        return self.ans

# Create TreeNode object from the input list
root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))

# Create an instance of the Solution class
solution = Solution()

# Call the function with the created TreeNode object
result = solution.diameterOfBinaryTree(root)

# Print the result
print(result)




'''
Given the root of a binary tree, return the length of the diameter of the tree.

The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

The length of a path between two nodes is represented by the number of edges between them.

 

Example 1:


Input: root = [1,2,3,4,5]
Output: 3
Explanation: 3 is the length of the path [4,2,1,3] or [5,2,1,3].
Example 2:

Input: root = [1,2]
Output: 1'''

File: JAVA/java/src/DiameterofBinaryTree.java
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class DiameterofBinayTree {
    private int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        depth(root);
        return ans;
    }

    private int depth(TreeNode p) {
        if (p == null) {
            return 0;
        }
        int left = depth(p.left);
        int right = depth(p.right);
        ans = Math.max(ans, left + right);
        return 1 + Math.max(left, right);
    }

    private TreeNode constructBinaryTree(Integer[] array) {
        if (array == null || array.length == 0) {
            return null;
        }

        TreeNode root = new TreeNode(array[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        for (int i = 1; i < array.length; i += 2) {
            TreeNode current = queue.poll();

            if (array[i] != null) {
                current.left = new TreeNode(array[i]);
                queue.offer(current.left);
            }

            if (i + 1 < array.length && array[i + 1] != null) {
                current.right = new TreeNode(array[i + 1]);
                queue.offer(current.right);
            }
        }

        return root;
    }

    public static void main(String[] args) {
        // Example input: [1, 2, 3, 4, 5]
        Integer[] array = {1, 2, 3, 4, 5};

        // Create an instance of DiameterOfBinaryTree class
        DiameterOfBinaryTree solution = new DiameterOfBinaryTree();

        // Create TreeNode object from the input array
        TreeNode root = solution.constructBinaryTree(array);

        // Call the method with the created TreeNode object
        int result = solution.diameterOfBinaryTree(root);

        // Print the result
        System.out.println(result);
    }
}
